e2e_tests/tests/cluster/test_master_restart.py

```diff
@@ -213,6 +214,102 @@ def test_master_restart_continued_experiment(
     assert "resources exited successfully with a zero exit code" in "".join(log.log for log in logs)
 
 
+def experiment_who_cancels_itself_then_waits(sess: api.Session) -> int:
+    return exp.create_experiment(
+        sess,
+        conf.fixtures_path("core_api/sleep.yaml"),
+        conf.fixtures_path("core_api"),
+        ["--config", "entrypoint='det e cancel $DET_EXPERIMENT_ID && sleep 500'"],
+    )
+
+
+@pytest.mark.managed_devcluster
+def test_master_restart_stopping(
+    restartable_managed_cluster: managed_cluster.ManagedCluster,
+) -> None:
+    _test_master_restart_stopping(restartable_managed_cluster)
+
+
+@pytest.mark.e2e_k8s
+def test_master_restart_stopping_k8s(
+    k8s_managed_cluster: managed_cluster_k8s.ManagedK8sCluster,
+) -> None:
+    _test_master_restart_stopping(k8s_managed_cluster)
+
+
+def _test_master_restart_stopping(managed_cluster_restarts: abstract_cluster.Cluster) -> None:
+    sess = api_utils.user_session()
+    sess._max_retries = urllib3.util.retry.Retry(total=5, backoff_factor=0.5)
+
+    exp_id = experiment_who_cancels_itself_then_waits(sess)
+    try:
+        exp.wait_for_experiment_state(sess, exp_id, bindings.experimentv1State.STOPPING_CANCELED)
+        managed_cluster_restarts.restart_master()
+
+        # Short wait so that we know it was killed by us and not preemption.
+        exp.wait_for_experiment_state(
+            sess, exp_id, bindings.experimentv1State.STOPPING_CANCELED, max_wait_secs=10
+        )
+    finally:
+        exp.kill_experiments(sess, [exp_id])
+        exp.wait_for_experiment_state(sess, exp_id, bindings.experimentv1State.CANCELED)
+
+    # All slots are empty, we don't leave a hanging container.
+    agentsResp = bindings.get_GetAgents(sess)
+    for a in agentsResp.agents:
+        if a.slots is not None:
+            for s in a.slots.values():
+                assert s.container is None, s.container.to_json()
+
+
+@pytest.mark.managed_devcluster
+def test_master_restart_stopping_ignore_preemption_still_gets_killed(
+    managed_cluster_restarts: managed_cluster.ManagedCluster,
+) -> None:
+    sess = api_utils.user_session()
+    sess._max_retries = urllib3.util.retry.Retry(total=5, backoff_factor=0.5)
+
+    exp_id = experiment_who_cancels_itself_then_waits(sess)
+    try:
+        exp.wait_for_experiment_state(sess, exp_id, bindings.experimentv1State.STOPPING_CANCELED)
+        managed_cluster_restarts.restart_master()
+        exp.wait_for_experiment_state(
```

Comment:
could check for 137 in this one? idk. just seems to more definitively assert what you are looking for.